File: leetcode/tree/889. Construct Binary Tree from Preorder and Postorder Traversal/js/main.js
// Time O(N^2)
// Space O(N)
const constructFromPrePost = (pre, post) => {
  let root = new TreeNode(pre[0]);
  let size = pre.length;
  let visited = Array(size).fill(false);

  visited[root.val] = true;

  helper([root]);

  return root;

  function helper(parent) {
    if (!parent.length) {
      return;
    }

    let children = [];

    for (let root of parent) {
      let left = null;
      let right = null;

      for (let i = 0; i < size; i++) {
        if (pre[i] == root.val && i < size) {
          left = pre[i + 1];
        }

        if (post[i] == root.val && i > 0) {
          right = post[i - 1];
        }
      }

      if (!visited[left] && left != null && left != undefined) {
        let node = new TreeNode(left);
        root.left = node;
        children.push(node);
        visited[left] = true;
      }

      if (!visited[right] && right != null && right != undefined) {
        let node = new TreeNode(right);
        root.right = node;
        children.push(node);
        visited[right] = true;
      }
    }

    helper(children);
  }
};

File: leetcode/tree/889. Construct Binary Tree from Preorder and Postorder Traversal/py/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    # Time O(N)
    # Space O(N)
    def constructFromPrePost(self, pre: List[int],
                             post: List[int]) -> TreeNode:
        # root-left-right
        # left-right-root

        root = TreeNode(pre[0])
        n = len(pre)

        # init the stack and the vistied state
        st = [root]
        visited = set([pre[0], None])

        mappingPre = {pre[n - 1]: None}
        mappingPost = {post[0]: None}

        for i in range(n - 1):
            mappingPre[pre[i]] = pre[i + 1]
            mappingPost[post[n - i - 1]] = post[n - i - 2]

        while st:
            node = st.pop()

            left = mappingPre[node.val]
            right = mappingPost[node.val]

            if left not in visited:
                node.left = TreeNode(left)
                st.append(node.left)
                visited.add(left)

            if right not in visited:
                node.right = TreeNode(right)
                st.append(node.right)
                visited.add(right)

        return root